compiler/ast/ast_query.nim

```diff
@@ -694,27 +694,34 @@ proc endsInNoReturn*(n: PNode): bool =
   ## Checks if expression `n` ends in an unstructured exit (raise, return,
   ## etc.) or a call of a noreturn proc. This is meant to be called on a
   ## semmed `n`.
-  var it = n
-  while it.kind in {nkStmtList, nkStmtListExpr} and it.len > 0 or
-        it.kind in {nkIfStmt, nkCaseStmt, nkBlockStmt, nkTryStmt} and it.typ.isEmptyType:
-    case it.kind
-    of nkStmtList, nkStmtListExpr, nkBlockStmt:
-      it = it.lastSon
-    of nkIfStmt, nkCaseStmt:
-      it = it.lastSon.lastSon
-    of nkTryStmt:
-      it =
-        case it[^1].kind
-        of nkFinally:
-          it[^2]
-        of nkExceptBranch:
-          it[^1]
-        of nkAllNodeKinds - {nkFinally, nkExceptBranch}:
-          unreachable()
-    else:
-      unreachable()
-  result = it.kind in nkLastBlockStmts or
-    it.kind in nkCallKinds and it[0].kind == nkSym and sfNoReturn in it[0].sym.flags
+  case n.kind
+  of nkStmtList, nkStmtListExpr:
+    result = n.len > 0 and endsInNoReturn(n[^1])
+  of nkBlockStmt, nkExceptBranch, nkElifBranch, nkElse, nkOfBranch:
+    result = endsInNoReturn(n[^1])
+  of nkIfStmt:
+    for it in n.items:
+      result = endsInNoReturn(it[^1])
+      if not result:
+        break
+  of nkCaseStmt:
+    # skip the selector expression
+    for i in 1..<n.len:
+      result = endsInNoReturn(n[i])
+      if not result:
+        break
```

Comment:
```suggestion
        break
    if result:
      # only noreturn if the 'case' is exhaustive
      result = n[^1].kind == nkElse
```

Reply:
Is this true, because if it's an enum and you list all possible values, then an `else` isn't required, i.e.:

```nim
var x: bool
case x
of true: discard
of false: discard
```

Reply:
You're right, those would lead to false negatives. An updated suggestion:
```suggestion
        break
    if result:
      result = n[0].typ.skipTypes(abstractRange).kind notin {tyString, tyFloat..tyFloat64} or
               n[^1].kind == nkElse
```
As per `semCase`, only `string` and `float` values can be non-exhaustive while omitting the `else` branch.

Reply:
I'll see, but I suspect I'm going to have to use/introduce a node flag to pass the information along.

Reply:
> You're right, those would lead to false negatives. An updated suggestion:
> 
> As per `semCase`, only `string` and `float` values can be non-exhaustive while omitting the `else` branch.

D'oh, I saw this after my change, this is simpler as I can rely on the initial exhaustiveness check instead of redoing the whole thing. I'll update my latest change with a simplified version.